## Логический тип данных и сравнения

In [ ]:
x = True
y = False
print(x)  # Выводит True
print(y)  # Выводит False


In [2]:
x = 5
y = 10
print(x == y)  # Выводит False
print(x < y)  # Выводит True


4


# Базовые операции

Для целого типа integers:
`+ - * / // % **`

для типа  floats:
`+ - * / **`

Boolean Логические:
* слова: `True` и `False` (!обратите внимание на заглавную букву)
* `==` равенство: `5 == 5` выдает `True`
* `!=` не равно: `5 != 5` выдает `False`
* `>` больше: `5 > 4` выдает `True`
* `>=`меньше: `5 >= 5` дает `True`
* Аналогично `<` и `<=`.

Логические операции:
* `and`, `or`, and `not`
* `True and False`
* `True or False`
* `not True`

(8%2==0) Четность

Ключевое слово None используется для определения нулевого значения или отсутствия значения . None не то же самое, что 0, False или пустая строка. None — это собственный тип данных (NoneType), и только None может быть None.

In [ ]:
x = None
if x is None:
    print("Переменная x не имеет значения")


# Управление памятью в Python

Управление памятью в Python — это процесс, в ходе которого компьютерная система резервирует часть или весь объём памяти для обработки и выполнения программ. Это позволяет прикладным программам считывать и записывать данные. Мы знаем, что объём памяти в компьютерной системе очень велик.


Необходимо выделить некоторое свободное пространство в памяти и сделать его доступным для выполнения программы. Такой способ выделения (то есть предоставления) памяти для выполнения программы называется выделением памяти.

Python автоматически управляет выделением и освобождением памяти во время выполнения программы. Нам не нужно выделять память при создании объектов или освобождать её при удалении объектов.

Интерпретатор Python отвечает за динамическое выделение и освобождение памяти. Другими словами, интерпретатор Python управляет памятью.

In [ ]:
x = 10

В Python всё считается объектом. Например, в Python объектами считаются числа, строки, функции, словари, списки, модули и пользовательские классы.
Мы можем получить к ним доступ с помощью идентификатора. Менеджер памяти Python выделяет память для объектов в соответствии с требованиями. Он использует для управления памятью частное пространство кучи, недоступное программисту.

Все объекты и структуры данных Python размещаются в этом пространстве кучи. Менеджер памяти с помощью функций Python API управляет этим частным пространством кучи для объектов Python.

В Python также есть встроенный сборщик мусора, который перерабатывает всю неиспользуемую память, чтобы в памяти кучи было свободное место.

In [1]:
x = 10
print(id(x))

2823743171152


![xkcd_python](https://github.com/trisha00001/IT-Career-Hub/blob/main/image/python-memory-management.png?raw=true)

Предположим, что мы присваиваем новое значение 10 новой переменной y. Посмотрите на следующий фрагмент кода.

Как вы можете видеть на выходе, обе ссылочные переменные x и y указывают на один и тот же целочисленный объект 10 из-за одинакового уникального идентификационного номера. 

In [ ]:
x = 10
print(id(x))
y = 10
print(id(y))

![xkcd_python](https://github.com/trisha00001/IT-Career-Hub/blob/main/image/same-id-value.png?raw=true)

Как вы можете видеть на выходе, обе ссылочные переменные x и y указывают на один и тот же целочисленный объект 10 из-за одинакового уникального идентификационного номера. 

Теперь рассмотрим следующий пример кода ниже.

In [ ]:
x = 10
print(id(x))

y = 10
print(id(y))

z = y
print(id(z))

Как видно из вывода, функция id() сгенерировала тот же идентификационный номер для целочисленного объекта 10. Следовательно, z также указывает на тот же объект.


![xkcd_python](https://github.com/trisha00001/IT-Career-Hub/blob/main/image/memory-management-in-Python.png?raw=true)

В этом примере функция id() генерирует уникальный идентификационный номер для объекта. Этот уникальный идентификационный номер представляет собой целочисленное значение, которое остаётся уникальным и постоянным для объекта Python в течение всего срока его существования, как показано на рисунке ниже. Идентификационный номер целочисленного объекта 10 равен 2823743171152

![xkcd_python](https://github.com/trisha00001/IT-Career-Hub/blob/main/image/referring-different-objects.png?raw=true)

In [ ]:
x = 10
print(id(x))
y = 10
print(id(y))

z = y
print(id(z))
x = x + 1
print(id(x))

Теперь, как вы можете видеть на выходе, функция id() сгенерировала другой идентификационный номер для этой операции x = x + 1. Следовательно, x не равно 11 и теперь относится к другому целочисленному объекту 11.

![xkcd_python](https://github.com/trisha00001/IT-Career-Hub/blob/main/image/03_01_02.gif?raw=true)

Сборка мусора - технология, позволяющая, с одной стороны, упростить программирование, избавив программиста от необходимости вручную удалять объекты, созданные в динамической памяти, с другой - устранить ошибки, вызванные неправильным ручным управлением памятью.

Алгоритм, используемый сборщиком мусора называется подсчетом ссылок. Python хранит журнал ссылок на каждый объект и автоматически уничтожает объект, как только на него больше нет ссылок.